include/aws/iotdevice/private/secure_tunneling_operations.h

```diff
@@ -0,0 +1,221 @@
+#include <aws/iotdevice/secure_tunneling.h>
+
+/*********************************************************************************************************************
+ * Operations
+ ********************************************************************************************************************/
+
+struct aws_secure_tunnel_operation;
+
+enum aws_secure_tunnel_operation_type {
+    AWS_STOT_NONE,
+    AWS_STOT_PING,
+    AWS_STOT_DATA,
+    AWS_STOT_STREAM_RESET,
+    AWS_STOT_STREAM_START
+};
+
+struct aws_secure_tunnel_message_storage {
+    struct aws_allocator *allocator;
+    struct aws_secure_tunnel_message_view storage_view;
+
+    bool ignorable;
+    int32_t stream_id;
+    struct aws_byte_cursor service_id;
+    struct aws_byte_cursor payload;
+
+    struct aws_byte_buf storage;
+};
+
+/* Basic vtable for all secure tunnel operations.  Implementations are per-message type */
+struct aws_secure_tunnel_operation_vtable {
+    void (*aws_secure_tunnel_operation_completion_fn)(
+        struct aws_secure_tunnel_operation *operation,
+        int error_code,
+        const void *completion_view);
+
+    /* Set the stream id of outgoing st_msg */
+    int (*aws_secure_tunnel_operation_set_stream_id_fn)(
+        struct aws_secure_tunnel_operation *operation,
+        struct aws_secure_tunnel *secure_tunnel);
+
+    /* Set the stream id of outgoing st_msg to +1 of the currently set stream id */
+    int (*aws_secure_tunnel_operation_set_next_stream_id_fn)(
+        struct aws_secure_tunnel_operation *operation,
+        struct aws_secure_tunnel *secure_tunnel);
+};
+
+/**
+ * This is the base structure for all secure tunnel operations.  It includes the type, a ref count, and list management.
+ */
+struct aws_secure_tunnel_operation {
+    const struct aws_secure_tunnel_operation_vtable *vtable;
+    struct aws_ref_count ref_count;
+    struct aws_linked_list_node node;
+
+    enum aws_secure_tunnel_operation_type operation_type;
+    const struct aws_secure_tunnel_message_view *message_view;
+
+    /* Size of the secure tunnel message this operation represents */
+    size_t message_size;
+
+    void *impl;
+};
+
+struct aws_secure_tunnel_operation_message {
+    struct aws_secure_tunnel_operation base;
+    struct aws_allocator *allocator;
+
+    struct aws_secure_tunnel_message_storage options_storage;
+};
+
+struct aws_secure_tunnel_operation_pingreq {
+    struct aws_secure_tunnel_operation base;
+    struct aws_allocator *allocator;
+};
+
+/*
+ * Secure tunnel configuration
+ */
+struct aws_secure_tunnel_options_storage {
+
+    // struct aws_secure_tunnel_options options;
+    struct aws_allocator *allocator;
+    struct aws_secure_tunnel *secure_tunnel;
+
+    /* backup */
+
+    struct aws_client_bootstrap *bootstrap;
+    struct aws_socket_options socket_options;
+    struct aws_http_proxy_options http_proxy_options;
+    struct aws_http_proxy_config *http_proxy_config;
+    struct aws_string *access_token;
```

Comment:
aws_string_destroy() doesn't expect const and I'm using it on cleanup. Another way to handle this could be to allocate an aws_byte_buf to keep the various strings in but I'm not seeing an immediately great reason for adding that complexity when aws_string is getting the job done. Anything changing these pointers should be fully controlled by us and they don't get modified outside of initial creation of the aws_string.